### PR TITLE
Fact generation for liveness calculations in Polonius

### PR DESCRIPTION
**Pull request by amandasystems**
This PR tracks ongoing work to extend `rustc` with support for generating variable use, definition, and later also drop output for the Polonius solver, the whole of which is being tracked in [Polonius Issue #104](https://github.com/rust-lang/polonius/issues/104).

### CODE REVIEW

**Review by nikomatsakis — APPROVED** [review]
Looks basically correct to me!

**Comment by nikomatsakis** [comment]
Marking as "waiting on author", as we're waiting on @albins to do some of the next steps.

File: src/librustc_mir/borrow_check/nll/type_check/liveness/polonius.rs
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/type_check/liveness/polonius.rs]: seems correct

**Comment by albins** [comment]
Now waiting for changes to land in Polonius: https://github.com/rust-lang/polonius/pull/105

**Comment by albins** [comment]
Hopefully not WIP anymore now that https://github.com/rust-lang/polonius/issues/104 is merged!

**Comment by albins** [comment]
We have decided to do preliminary work on integrating initializiation tracking as well before merging this in order to get (hopefully) exactly the same output as rustc from Polonius, so back to WIP!

**Comment by albins** [comment]
We have decided to do preliminary work on integrating initializiation tracking as well before merging this in order to get (hopefully) exactly the same output as rustc from Polonius, so back to WIP!

**Comment by albins** [comment]
This branch now has preliminary liveness calculations that uses a hack to extract initialisation status, which is needed for correct drop-liveness calculations. When we have initialisation tracking in Polonius that hack will go away, but for now and to the best of our knowledge these inputs compute exactly the same region liveness tuples as rustc previously generated, including for tricky edge cases™.

**Review by nikomatsakis — APPROVED** [review]

**Comment by lqd** [comment]
Just in case bors ignores commands on S-waiting-on-author:

@bors r=nikomatsakis

**Comment by lqd** [comment]
Rebase is done

@bors r=nikomatsakis

**Pull request merged**
